### ISSUE TITLE
Pyinstaller interaction with cryptography module.

### ISSUE THREAD

**Issue by TheOtherGuy** [pyinstaller/pyinstaller#2013]
Is `Pyinstaller` working with the `cryptography` module as of now ?

https://github.com/pyinstaller/pyinstaller/issues/1425
https://github.com/pyca/cryptography/issues/2039

I am using `Pyinstaller 3.2` ,`Python 3.4.4` , `Windows 7 32 bit`.
The executable I created using `Pyinstaller` fails when with this error `ValueError: multibackend cannot be initialised with no backends`. From some searching, this is related to the `cryptography` module.

Full stack trace:

```
Unknown exception: Multibackend cannot be initialized with no backends. If you are seeing this error when trying to use default_backend() please try uninstalling and reinstalling cryptography.
Traceback (most recent call last):
  File "site-packages\paramiko\transport.py", line 1757, in run
  File "site-packages\paramiko\kex_group1.py", line 75, in parse_next
  File "site-packages\paramiko\kex_group1.py", line 111, in _parse_kexdh_reply
  File "site-packages\paramiko\transport.py", line 1602, in _verify_key
  File "site-packages\paramiko\rsakey.py", line 58, in __init__
  File "site-packages\cryptography\hazmat\backends\__init__.py", line 35, in default_backend
  File "site-packages\cryptography\hazmat\backends\multibackend.py", line 33, in __init__
ValueError: Multibackend cannot be initialized with no backends. If you are seeing this error when trying to use default_backend() please try uninstalling and reinstalling cryptography.

Traceback (most recent call last):
  File "automa.py", line 383, in <module>
  File "automa.py", line 150, in sshconn
  File "site-packages\netmiko\ssh_dispatcher.py", line 84, in ConnectHandler
  File "site-packages\netmiko\base_connection.py", line 68, in __init__
  File "site-packages\netmiko\base_connection.py", line 165, in establish_connection
  File "site-packages\paramiko\client.py", line 338, in connect
  File "site-packages\paramiko\transport.py", line 493, in start_client
  File "site-packages\paramiko\transport.py", line 1757, in run
  File "site-packages\paramiko\kex_group1.py", line 75, in parse_next
  File "site-packages\paramiko\kex_group1.py", line 111, in _parse_kexdh_reply
  File "site-packages\paramiko\transport.py", line 1602, in _verify_key
  File "site-packages\paramiko\rsakey.py", line 58, in __init__
  File "site-packages\cryptography\hazmat\backends\__init__.py", line 35, in default_backend
  File "site-packages\cryptography\hazmat\backends\multibackend.py", line 33, in __init__
ValueError: Multibackend cannot be initialized with no backends. If you are seeing this error when trying to use default_backend() please try uninstalling and reinstalling cryptography.
Failed to execute script automa
```

Pyinstaller seems to have fixed the issue in this script, but the executable still faces issues.
[PyInstaller/hooks/hook-cryptography.py](https://github.com/pyinstaller/pyinstaller/blob/master/PyInstaller/hooks/hook-cryptography.py)

Based on the output of `archive_viewer.py`, it appears that the `cryptography` module is already packaged into the executable.

Truncated 1st level output:

```
 (458112, 57624, 119296, 1, 'b', '_cffi_backend.pyd'),
 (515736, 38697, 85504, 1, 'b', '_ctypes.pyd'),
 (554433, 77948, 182272, 1, 'b', '_decimal.pyd'),
 (632381, 446240, 922112, 1, 'b', '_hashlib.pyd'),
 (1078621, 74080, 137216, 1, 'b', '_lzma.pyd'),
 (1152701, 6129, 12288, 1, 'b', '_multiprocessing.pyd'),
 (1158830, 21467, 47104, 1, 'b', '_socket.pyd'),
 (1180297, 650025, 1331200, 1, 'b', '_ssl.pyd'),
 (1830322,
  753828,
  1819136,
  1,
  'b',
  'cryptography.hazmat.bindings._openssl.pyd'),
 (2584150, 391, 712, 1, 'b', 'automa.exe.manifest'),
 (2584541, 59293, 133120, 1, 'b', 'pyexpat.pyd'),
 (2643834, 1223183, 2744832, 1, 'b', 'python34.dll'),
 (3867017, 40591, 104960, 1, 'b', 'pywintypes34.dll'),
 (3907608, 5054, 9728, 1, 'b', 'select.pyd'),
 (3912662, 294033, 758784, 1, 'b', 'unicodedata.pyd'),
 (4206695, 41554, 108032, 1, 'b', 'win32security.pyd'),
```

Truncated output of out00-PYZ.pyz:

```
 ? O out00-PYZ.pyz
 Name: (ispkg, pos, len)
 'cryptography': (1, 108606, 553),
 'cryptography.__about__': (0, 109159, 513),
 'cryptography.exceptions': (0, 109672, 1011),
 'cryptography.hazmat': (1, 110683, 345),
 'cryptography.hazmat.backends': (1, 111028, 535),
 'cryptography.hazmat.backends.interfaces': (0, 111563, 3112),
 'cryptography.hazmat.backends.multibackend': (0, 114675, 3638),
 'cryptography.hazmat.bindings': (1, 118313, 195),
 'cryptography.hazmat.bindings.openssl': (1, 118508, 198),
 'cryptography.hazmat.bindings.openssl._conditional': (0, 118706, 4025),
 'cryptography.hazmat.bindings.openssl.binding': (0, 122731, 3483),
 'cryptography.hazmat.primitives': (1, 126214, 191),
 'cryptography.hazmat.primitives.asymmetric': (1, 126405, 706),
 'cryptography.hazmat.primitives.asymmetric.dsa': (0, 127111, 2700),
 'cryptography.hazmat.primitives.asymmetric.ec': (0, 129811, 3686),
 'cryptography.hazmat.primitives.asymmetric.padding': (0, 133497, 1200),
 'cryptography.hazmat.primitives.asymmetric.rsa': (0, 134697, 3976),
 'cryptography.hazmat.primitives.asymmetric.utils': (0, 138673, 1220),
 'cryptography.hazmat.primitives.ciphers': (1, 139893, 352),
 'cryptography.hazmat.primitives.ciphers.algorithms': (0, 140245, 1466),
 'cryptography.hazmat.primitives.ciphers.base': (0, 141711, 2710),
 'cryptography.hazmat.primitives.ciphers.modes': (0, 144421, 2029),
 'cryptography.hazmat.primitives.hashes': (0, 146450, 1997),
 'cryptography.hazmat.primitives.serialization': (0, 148447, 2587),
 'cryptography.utils': (0, 151034, 2203),
```

I tried compiling with both -F (single file) and the default directory based format. Both resulted in the same errors.

**Comment by TheOtherGuy** [comment]
Hi, is there anything I can possibly do to help solve this issue ? 
Thanks & Regards.

**Comment by six8** [comment]
Having the same issue on Linux. I applied the monkey patch from https://github.com/pyca/cryptography/issues/2039 to get it to work.

```
def patch_crypto_be_discovery():

    """
    Monkey patches cryptography's backend detection.
    Objective: support pyinstaller freezing.
    """

    from cryptography.hazmat import backends

    try:
        from cryptography.hazmat.backends.commoncrypto.backend import backend as be_cc
    except ImportError:
        be_cc = None

    try:
        from cryptography.hazmat.backends.openssl.backend import backend as be_ossl
    except ImportError:
        be_ossl = None

    backends._available_backends_list = [
        be for be in (be_cc, be_ossl) if be is not None
    ]

```

**Comment by TheOtherGuy** [comment]
I can confirm on my end that the above monkey patching works for my setup. I am still hoping for a more permanent solution though.

**Comment by IWouldLikeToKnow** [comment]
@six8, @TheOtherGuy 

Hello,

I'm facing the same issue, but being quite new to programming I have no idea on how to apply this patch.

Would one of you possibly help me with this, please? It would be greatly appreciated,

**Comment by stephenrauch** [comment]
The monkey patch code presented above needs to be run before the backend is selected (or auto-seletected?).  Some place in your script, presumably near the beginning, the backend is loaded/selected/etc... (I don't know this module, so don't know what terms they use). Before that happens, call the monkey patch function

**Comment by codewarrior0** [comment]
Cryptography finds its backends via the `pkg_resources.iter_entry_points` function. This function searches a list of Distributions that is built earlier by searching the folders on `sys.path` for files and/or folders named with `.egg, .egg-info, .egg-link, .dist-info` - these files contain the setuptools metadata for the installed Distribution including its list of entry points. Maybe the solution is as simple as including the `.egg-info` folder for each of the Cryptography backends.

**Comment by IWouldLikeToKnow** [comment]
@stephenrauch @codewarrior0 

Thank you very much for your help.

**Comment by TheOtherGuy** [comment]
@IWouldLikeToKnow 

You can apply the monkey patch by pasting the `patch_crypto_be_discovery()` function into your python script, and then executing the function somewhere around the beginning of your script by inserting a line  `patch_crypto_be_discovery()`.

**Comment by IWouldLikeToKnow** [comment]
@TheOtherGuy 

It worked!

Thank you very much to all of you.

**Comment by adityaDotmappers** [comment]
It Worked !!!!
Thank you all

**Comment by Rahushi** [comment]
I am new to Python and facing same issue. I have used 'paramiko' package in my script which works absolutely fine on console. But when I create executable using py2exe or cxFreeze, it throws following error. I tried the above mentioned 'patch_crypto_be_discovery' solution but still getting same error. I have been trying hard to make it work with no luck. Any help will be much appreciated.

`Traceback (most recent call last):
  File "gui.py", line 113, in OnRun
  File "paramiko\client.pyc", line 338, in connect
  File "paramiko\transport.pyc", line 493, in start_client
ValueError: Multibackend cannot be initialized with no backends. If you are seeing this error when trying to use default_backend() please try uninstalling and reinstalling cryptography.`

**Comment by six8** [comment]
Along with the `patch_crypto_be_discovery` monkeypatch, I use the following build options for pyinstaller:

```
        --hidden-import cryptography.hazmat.backends.openssl \
        --hidden-import cffi \
```

**Comment by Rahushi** [comment]
@six8 : Thanks for the reply. I am using py2exe. Could you please tell me how to create using pyinstaller. I do not see any error related to those modules though.

**Comment by Rahushi** [comment]
@six8 
I did add 'cffi' and it worked perfectly fine now using cxFreeze. Thankyou very much!

**Comment by Russel777** [comment]
Hey, I have the same problem with cx_freeze - solution is the same like "six8 commented on 23 Sep":
1) use the "patch_crypto_be_discovery" and call it before you connect to server
2) when you build app, use next: 
python setup.py build_exe --includes cryptography.hazmat.backends.openssl,cffi

**Comment by lbleon** [comment]
@Russel777 
I have the same problem with cx_freeze. And it's worked with your solution. Thanks a lot:)

**Comment by aristotle23** [comment]
I tried the patch_crypto_be_discover
`file "site-package\cryptotagraphy\hazmat\backends\multibackend.py" line 31 in __init__ 
TypeError: object of type "NoneType" has no len`

**Comment by dessant** [comment]
This should be fixed, see https://github.com/pyca/cryptography/issues/2039#issuecomment-264857116.

**Comment by xoviat** [comment]
@htgoebel Closed in c91952d1cbe2b082b1ba895f05022ecc47c03cd3.

**Comment by liashchynskyi** [comment]
@Russel777 thanks)) it works fine)

**Comment by DaniuLi** [comment]
This should be fixed, but It does not worked ...(cryptography2.2.2, Pyinstaller 3.3.1 , Python 3.6.5 , Linux)

why?

stack trace:

<pre>
8625 INFO: Loading module hook "hook-cryptography.py"...
Traceback (most recent call last):
  File "/usr/local/bin/pyinstaller", line 11, in <module>
    load_entry_point('PyInstaller==3.3.1', 'console_scripts', 'pyinstaller')()
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/__main__.py", line 94, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/__main__.py", line 46, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/build_main.py", line 791, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/build_main.py", line 737, in build
    exec(text, spec_namespace)
  File "<string>", line 16, in <module>
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/build_main.py", line 213, in __init__
    self.__postinit__()
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/datastruct.py", line 161, in __postinit__
    self.assemble()
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/build_main.py", line 472, in assemble
    module_hook.post_graph()
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/imphook.py", line 410, in post_graph
    self._load_hook_module()
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/building/imphook.py", line 377, in _load_hook_module
    self.hook_module_name, self.hook_filename)
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/compat.py", line 744, in importlib_load_source
    return mod_loader.load_module()
  File "<frozen importlib._bootstrap_external>", line 399, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 823, in load_module
  File "<frozen importlib._bootstrap_external>", line 682, in load_module
  File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/hooks/hook-cryptography.py", line 23, in <module>
    datas = copy_metadata('cryptography')
  File "/usr/local/lib/python3.6/site-packages/PyInstaller/utils/hooks/__init__.py", line 852, in copy_metadata
    dist = pkg_resources.get_distribution(package_name)
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 472, in get_distribution
    dist = get_provider(dist)
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 344, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'cryptography' distribution was not found and is required by the application
</pre>

**Comment by nickvans** [comment]
For what it's worth, I had this problem with Python 3.6, and PyInstaller 4.4 in July 2021. I tried the monkey patch but still had import errors. The thing that worked for me was to include the dlls for 'libcrypto-1_1-x64.dll' and 'libssl-1_1-x64.dll' which I got out of <anaconda env>\Library\bin directory. In my case I copied these dlls into my app dir (so other people on my team could create the exe using pyinstaller from our git repo) and noted them in the PyInstaller .spec file like so

```
a = Analysis(['Main.py'],
             ...
             datas=[
                ('libcrypto-1_1-x64.dll', '.'),
                ('libssl-1_1-x64.dll', '.')
                ],
            ...)
```

I didn't need the monkey patch or any hidden imports once I did this.
